Allow password change with identity verification

bl-nero opened the pull request:
Contributes to https://github.com/gravitational/teleport/issues/36233
Partially implements [RFD 0159](https://github.com/gravitational/teleport/blob/master/rfd/0159-password-status.md).

This is a backend change that supports the related PR https://github.com/gravitational/teleport/pull/38728.

bl-nero committed:
Allow password change with identity verificaiton

zmb3 commented:
/excludeflake *

codingllama commented on lib/auth/auth_login_test.go:
```diff
@@ -1187,10 +1187,14 @@ func configureForMFA(t *testing.T, srv *TestTLSServer) *configureMFAResp {
 	webDev, err := RegisterTestDevice(ctx, clt, "web-1", proto.DeviceType_DEVICE_TYPE_WEBAUTHN, totpDev)
 	require.NoError(t, err)
 
+	passwordlessDev, err := RegisterTestDevice(ctx, clt, "passwordless-1", proto.DeviceType_DEVICE_TYPE_WEBAUTHN, totpDev, WithPasswordless())
```
If this is only for a single test consider moving the device creation there, otherwise every caller of this helper will have to pay the costs of an additional ceremony.

codingllama commented on lib/auth/methods.go:
```diff
@@ -260,7 +262,7 @@ var (
 	// authenticateWebauthnError is the generic error returned for failed WebAuthn
 	// authentication attempts.
-	authenticateWebauthnError = trace.AccessDenied("invalid Webauthn response")
+	authenticateWebauthnError = trace.AccessDenied("invalid credentials")
```
Could we backport the message changes? Do you think that's worth it?

bl-nero replied:
Yes. I was planning to backport this change along with the UI, but since the UI code in this PR has been updated to be compatible with the new flow, I think I can equally well backport it separately.

codingllama commented on lib/auth/methods.go:
```diff
@@ -295,8 +302,9 @@ type verifyMFADeviceLocksParams struct {
 func (a *Server) authenticateUser(
 	ctx context.Context,
 	req AuthenticateUserRequest,
+	requiredScope mfav1.ChallengeScope,
```
Should we pass the entire extensions object for future proofing?

codingllama commented on lib/auth/methods.go:
```diff
@@ -341,7 +349,11 @@ func (a *Server) authenticateUser(
+func (a *Server) authenticateUserInternal(
+	ctx context.Context,
+	req AuthenticateUserRequest,
+	requiredScope mfav1.ChallengeScope,
```
Ditto for passing the entire extensions object.

codingllama commented on lib/auth/methods.go:
```diff
@@ -382,7 +394,13 @@ func (a *Server) authenticateUserInternal(ctx context.Context, req AuthenticateU
+			// old password, we need to require identity verification (i.e. a resident
+			// token used for MFA).
+			if requiredScope == mfav1.ChallengeScope_CHALLENGE_SCOPE_CHANGE_PASSWORD && req.Pass == nil {
+				requiredExt.UserVerificationRequirement = string(protocol.VerificationRequired)
```
Let's move this to ChangePassword, I think it's better placed in a "higher" layer. `authenticateUserInternal` shouldn't be concerned with the specifics of changing passwords.

codingllama commented on lib/auth/methods.go:
```diff
@@ -341,7 +349,11 @@ func (a *Server) authenticateUser(
+) (mfaDev *types.MFADevice, user string, err error) {
 	if err := req.CheckAndSetDefaults(); err != nil {
 		return nil, "", trace.Wrap(err)
 	}
```
Could we double-check the scope in the passwordless branch and error if it's wrong?

bl-nero replied:
It's already there, in the `authenticatePasswordless` function that doesn't even take the required extensions/scope and just substitutes its own. I can double check here, though.

bl-nero replied:
Ok, hold on. That doesn't really make any sense to double-check a parameter that ends up not getting used. @codingllama  Do you mean to pass the required extensions down to `authenticatePasswordless`, and validate it there instead of creating a new one?

codingllama replied:
I meant something like this:

```go
	// Only one path if passwordless, other variants shouldn't see an empty user.
	if passwordless {
		if requiredScope != mfav1.ChallengeScope_CHALLENGE_SCOPE_LOGIN {
			return nil, "", trace.BadParameter("passwordless can only be used with the PASSWORDLESS scope")
		}
		return a.authenticatePasswordless(ctx, req)
	}
```

This is just so we flag any eventual programming errors or mismatch.

codingllama replied:
Ping on this one? I think it's only comment left.

bl-nero replied:
Strange. It was there. Perhaps I accidentally hit the undid it.

codingllama commented on lib/web/mfa.go:
```diff
@@ -132,9 +133,10 @@ func (h *Handler) addMFADeviceHandle(w http.ResponseWriter, r *http.Request, par
+	IsMFARequiredRequest        *isMFARequiredRequest                 `json:"is_mfa_required_req"`
+	ChallengeScope              int                                   `json:"challenge_scope"`
+	ChallengeAllowReuse         bool                                  `json:"challenge_allow_reuse"`
+	UserVerificationRequirement *protocol.UserVerificationRequirement `json:"user_verification_requirement"`
```
```suggestion
	UserVerificationRequirement string `json:"user_verification_requirement"`
```

Go won't magically coerce the unmarshaled values to the enum set, so this is just giving us a false sense of (type) safety. Plus, this is just a string on ChallengeExtensions.

bl-nero replied:
I can change it, though I'd like to point out that we already have some JSON-serializable structures that have fields of this type explicitly typed.

codingllama replied:
Yes, please change this.

What other structs do we have?

bl-nero replied:
OK, will do.

Other structs:
https://github.com/gravitational/teleport/blob/master/lib/auth/webauthntypes/webauthn.go#L46
https://github.com/gravitational/teleport/blob/master/lib/auth/webauthntypes/webauthn.go#L193

codingllama replied:
Structs within webauthntypes are fine.

codingllama commented on lib/web/mfa.go:
```diff
@@ -163,14 +165,20 @@ func (h *Handler) createAuthenticateChallengeHandle(w http.ResponseWriter, r *ht
 		allowReuse = mfav1.ChallengeAllowReuse_CHALLENGE_ALLOW_REUSE_YES
 	}
 
+	uvr := "" // default
+	if req.UserVerificationRequirement != nil {
+		uvr = string(*req.UserVerificationRequirement)
+	}
```
No need to do any of this, just pass the plain string along.

codingllama commented on lib/web/password.go:
```diff
@@ -39,7 +39,7 @@ type changePasswordReq struct {
 	SecondFactorToken string `json:"second_factor_token"`
 	// WebauthnAssertionResponse is a Webauthn response
-	WebauthnAssertionResponse *wantypes.CredentialAssertionResponse `json:"webauthnAssertionResponse"`
+	WebauthnAssertionResponse *wantypes.CredentialAssertionResponse `json:"webauthn_assertion_response"`
```
Why the tag change here? This is a breaking change.

bl-nero replied:
I just spotted an inconsistency and thought it's not a breaking change if it only affects the web API, which is being updated in the same PR.

codingllama replied:
Discussed out-of-band, whatever you prefer is fine.

avatus replied:
This is something we need to address eventually with the web team. This is a breaking change if our web client hits a proxy of a different version. As of now, we haven't seen too many of these issues, but when customers who run multiple proxies in a load balancer, this will break (or leaf clusters too i think?). It isn't that frequent, but frequent enough that we have a quarter goal this quarter to be able to serve the web asset bundle to different versions.

I would suggest not including this line

zmb3 replied:
+1 please avoid the breaking change.

codingllama commented on web/packages/teleport/src/Account/ChangePassword/useChangePassword.ts:
```diff
@@ -18,15 +18,37 @@
 
 import authService from 'teleport/services/auth';
 import cfg from 'teleport/config';
+import { MfaChallengeScope } from 'teleport/services/auth/auth';
```
Should we move these to #38728 ?

codingllama replied:
Fair enough.

codingllama commented on lib/auth/password_test.go:
```diff
@@ -237,37 +239,54 @@ func TestServer_ChangePassword(t *testing.T) {
+			name:                        "OK TOTP-based change",
+			oldPass:                     password,
+			newPass:                     "llamasarecool11",
+			userVerificationRequirement: "discouraged",
```
```suggestion
```

So it's more consistent with prod usage, which won't be setting this?

codingllama commented on lib/auth/password_test.go:
```diff
@@ -237,37 +239,54 @@ func TestServer_ChangePassword(t *testing.T) {
+			name:                        "OK Webauthn-based change",
+			oldPass:                     "llamasarecool11",
+			newPass:                     "llamasarecool13",
+			userVerificationRequirement: "discouraged",
```
```suggestion
```

Same. If you want to explicitly check "discouraged" I suggest adding a new test scenario, distinct from the existing ones.

codingllama commented on lib/auth/password_test.go:
```diff
@@ -237,37 +239,54 @@ func TestServer_ChangePassword(t *testing.T) {
-	ctx := context.Background()
+	ctx := authz.ContextWithUser(
+		context.Background(),
+		authz.LocalUser{Username: username, Identity: tlsca.Identity{Username: username}},
```
I would rather use a client for this user than mess with the context's internal variables.

Try this:

```go
	userClient, err := srv.NewClient(TestUser(username))
```

and then use `userClient` for the user->Auth interactions.

bl-nero replied:
Yeah, will do. I thought that there's something important in the fact that we're using the server directly, since the test's name mentions the word "server" explicitly.

codingllama commented on lib/auth/password_test.go:
```diff
@@ -237,37 +239,54 @@ func TestServer_ChangePassword(t *testing.T) {
-						Username: username,
-						Password: oldPass,
-					},
+				Request: &proto.CreateAuthenticateChallengeRequest_ContextUser{
+					ContextUser: &proto.ContextUser{},
+				},
```
Are we changing the UI to do this as well?

I wonder if we should push this part of the request to the test table too, so we don't change the character of the tests.

bl-nero replied:
Yes, the client code changes (see `useChangePassword.ts`), and the "old way" goes away. The change there is minimal, just for compatibility sake, and the "big" change comes with the frontend PR. Unless we support third-party web clients (which, AFAIK, is not the case), I'd say we should change the test behavior here.

codingllama replied:
I think we should keep at least one test with the old behavior, given that the functionality is still there.

codingllama commented on lib/auth/password_test.go:
```diff
@@ -307,42 +324,74 @@ func TestServer_ChangePassword_FailsWithoutOldPassword(t *testing.T) {
-			ContextUser: &proto.ContextUser{},
+	tests := []struct {
+		name                        string
+		userVerificationRequirement string
```
Drop this and let the string be empty in the request? It should be more reflective of production usage.

```suggestion
```

codingllama commented on lib/auth/password_test.go:
```diff
@@ -307,42 +324,74 @@ func TestServer_ChangePassword_FailsWithoutOldPassword(t *testing.T) {
-
+	ctx := authz.ContextWithUser(
+		context.Background(),
+		authz.LocalUser{Username: mfa.User, Identity: tlsca.Identity{Username: mfa.User}},
```
This shouldn't be necessary as we are using a client with TestUser below.

codingllama reviewed (commented):
Thanks for splitting the backend changes!

bl-nero committed:
Review

bl-nero commented on web/packages/teleport/src/services/auth/types.ts:
```diff
@@ -70,3 +72,17 @@ export type ResetPasswordWithWebauthnReqWithEvent = {
+  scope: MfaChallengeScope;
+  allowReuse?: boolean;
+  isMfaRequiredRequest?: IsMfaRequiredRequest;
+  userVerificationRequirement?: UserVerificationRequirement;
```
@codingllama If you don't insist, I'd like to keep it as is, since it's idiomatic on the frontend side to have strongly typed fields that lack proper validation on deserialization. And here, in this case, we are talking about a type that won't even be deserialized, just serialized; here, the benefit of specifying type explicitly is having the code self-documented and checked internally.

codingllama replied:
Sure, I'm not really a frontend reviewer, so it's your call.

bl-nero committed:
Revert renaming the webauthnAssertionResponse field

codingllama commented on lib/auth/methods.go:
```diff
@@ -382,8 +403,7 @@ func (a *Server) authenticateUserInternal(ctx context.Context, req AuthenticateU
 					Webauthn: wantypes.CredentialAssertionResponseToProto(req.Webauthn),
```
Should we tighten the checks above to make sure a nil `req.Pass` won't just slip by? Around line 375, aka `if req.Pass != nil {`.

```go
			if req.Pass == nil && requiredExt.UserVerificationRequirement != string(protocol.UserVerificationRequirement) {
				return nil, trace.BadParameter("password required")
			}
```

codingllama replied:
Nevermind, I see this can't be true for all situations.

bl-nero committed:
review

codingllama commented on lib/auth/password_test.go:
```diff
@@ -235,41 +235,101 @@ func TestServer_ChangePassword(t *testing.T) {
+			oldPass:          "llamasarecool12",
+			newPass:          "llamasarecool13",
+			device:           mfa.WebDev,
+			challengeRequest: &proto.CreateAuthenticateChallengeRequest{},
```
Incomplete request?

bl-nero replied:
@codingllama Holy cow. Not only an incomplete request, but also incidentally, a piece of test that shouldn't have passed. I have noticed that we are actually falling here into a case where `loginFlow.finish()` uses a backward-compatible logic where it allows UNSPECIFIED scope, and this challenge request passes scope validation, even though it shouldn't. I'll definitely add a test that verifies that we can't do it and will provide a special case inside a special case there in `loginFlow.finish()` to treat the password change scope more strictly.

(As for backwards compatibility, I don't think we should care here, since only the web client offers password change ability anyway.)

codingllama replied:
Gotcha, thanks for the careful testing here. This could cause problems if Proxy and Auth versions are mismatched, which is something to note, but that's likely OK given how serious password changes are.

bl-nero committed:
Fix a bug, add a test

bl-nero commented:
@codingllama I fixed a bug that I discovered when reading your comments on the unit test, PTAL. I added a couple of failure cases to the unit tests and retested everything right after the push:
- Changing password with a passwordless MFA
- Changing password with a YubiKey
- Changing password with OTP device
- Changing password without confirmation (user doesn't have any auth device).
All with both success and failure cases.
Note that I didn't retest manually the passwordless case _without_ the old password, I'll do it on the UI branch later today.

bl-nero commented:
Alright, from the unit test run, looks like I broke device locking this time. Fixing.

bl-nero committed:
Fix a broken test

bl-nero commented:
Aaaand fixed. It was just a test that wasn't updated for the new scope requirement.
